lib/sles4sap_publiccloud.pm

```diff
@@ -110,6 +111,59 @@ sub get_promoted_hostname {
     return join("", @master);
 }
 
+=head2 sles4sap_cleanup
+
+    Clean up Network peering and qesap deployment
+
+=cut
+
+sub sles4sap_cleanup {
+    my ($self, %args) = @_;
+
+    record_info('Cleanup',
+        join(' ',
+            'cleanup_called:', $args{cleanup_called} // 'undefined',
+            'network_peering_present:', $args{network_peering_present} // 'undefined',
+            'ansible_present:', $args{ansible_present} // 'undefined'));
+
+    qesap_upload_logs();
+    if ($args{network_peering_present}) {
+        delete_network_peering();
```

Comment:
why you want to do it even if it was done ? I mean why you want to do it two times ?

Reply:
The idea is, if a peering is present (we set this var in our network_peering module), delete the peering before deleting the infrastructure. It's not done twice - this is the only place where the peering gets deleted when there's a failure in a module that inherits your (PC) basetest, instead of our own basetest class.

Reply:
there is very good reason why we have `cleanup_called` variable . I can not clearly describe you whole picture now out of the blue but I can assure you that all the code which exists in `post_fail_hook` /  `post_run_hook` and not protected by `unless $self->{cleanup_called};` WILL be called two times

Reply:
Oh, now I got it. You mean that the peering cleanup is before `return 0 if ($args{cleanup_called});`. However this does not matter in this case, as if the peering is already destroyed,the peering destruction function won't do anything. 

As it's harmless, I'm placing it above the check to make sure that peering cleanup will still be called if for some reason resource cleanup has already taken place (the peering from IBSM->deleted_resource will still be present in that case).

Reply:
> As it's harmless

there are 2 calls of azure cli and second one does not have proceed on failure so in case of some glitch in Azure API  may unreasonably fail execution flow , also there is a call to scale timeout :thinking:  not sure

Reply:
this discussion does not make sense anymore . after you moved call inside `_cleanup` whole function won't be called at all second time . **our** code will take care about that :)